### DISCUSSION TITLE
How to explode TEXT using SHX fonts

### DISCUSSION THREAD

**Discussion by typhoon71** [mozman/ezdxf#839] [Q&A]
I'm trying to get a TEXT converted to ARC, LINEs, ... to mill it; this is the source text:
![immagine](https://user-images.githubusercontent.com/1866248/220989014-493af6bf-5829-41f5-954c-23d89ebaca07.png)

I see text2path does convert text to paths, and I can get this:

![immagine](https://user-images.githubusercontent.com/1866248/220989073-a98e554d-473e-4092-b067-1c9057161ac8.png)

This is different from what I need:

![immagine](https://user-images.githubusercontent.com/1866248/220989711-2d118db7-2e06-4bb5-9288-1d5f018c3df2.png)

This seems the same of the first image, but it's actually the result of exploding the text with AutoCAD express tools (or LibreCAD explode tool): those are arcs, lines, polilynes.
Basically I don't want the path around the font, but inside - obviously I'm not after fancy fonts as you can see.

Is this doable with text2path?

**Comment by mozman** [comment]
tl;dr: it's not possible yet, but there are experimental features for it

Shapes and vector fonts in AutoCAD are stored in SHX files and the source code for SHX
files is a SHP file. (not the GIS file format!)

The `ezdxf` add-on `text2path` uses TTF replacement fonts to render SHX fonts, these fonts
always produce outlines as you showed.  I have already implemented a SHX/SHP font file
parser and rendering support in the module `ezdxf.shapefile`, but big-fonts are not
supported yet, which supports characters beyond ASCII.

The big problem: Copyright

I can not include the SHX/SHP fonts in `ezdxf`.

If you have any Autodesk product installed you can add the font directory to the support
directories of `ezdxf`:

    ezdxf.options.support_dirs = ["C:\Program Files\Autodesk\DWG TrueView 2022 - English\Fonts"]

If you don't have any CAD software which includes SHX fonts, you have to search the
internet!

LibreCAD circumvents this problem by using other replacement fonts with the ".lff"
extension, support for these font files is already on my TODO list.

An example for using SHX fonts can be found here: [examples/render/shapefile_rendering.py](https://github.com/mozman/ezdxf/blob/master/examples/render/shapefile_rendering.py)

The following example renders the text "7410/A" by the "ISO.SHX" font:

```Python
import pathlib
import ezdxf
from ezdxf import shapefile, path, zoom
from ezdxf.filemanagement import find_support_file
from ezdxf.document import Drawing

CWD = pathlib.Path("~/Desktop/Outbox").expanduser()
if not CWD.exists():
    CWD = pathlib.Path(".")

# Add the directory containing the shx/shp files to your config file:
# 1. create a default config file in home directory at ~/.config/ezdxf/ezdxf.ini
#
#     $ ezdxf config --home
#
# 2. add the font directory as support dir, this is my config file:
# [core]
# default_dimension_text_style = OpenSansCondensed-Light
# test_files = ~/src/dxftest
# font_cache_directory =
# support_dirs = ~/src/shx-fonts
# 	~/src/ctb
#
# or
#
# set the support directories in the script:
# ezdxf.options.support_dirs = ["/absolut/path/to/shx/files"]


def render_txt(fontname: str, text: str) -> Drawing:
    doc = ezdxf.new()
    msp = doc.modelspace()
    fontname = find_support_file(fontname, ezdxf.options.support_dirs)
    # fontname should be the file path to the SHX font
    font = shapefile.readfile(fontname)
    # text_path is a ezdxf.path.Path instance!
    text_path = font.render_text(text)
    path.render_splines_and_polylines(msp, [text_path])
    return doc


doc = render_txt("iso.shx", "7410/A")
zoom.extents(doc.modelspace())
doc.saveas(CWD / "iso_shx.dxf")
```

If everything works, the result should look like this:

![image](https://user-images.githubusercontent.com/4230966/221130209-94c0770f-5478-41be-8151-1d7ebc9768d9.png)

Additional tips:

The text style is stored in DXF attribute "style" of the TEXT entity and the font name is stored in 
the text style attribute "font", the font name may not include the `.shx` extension like the default font "txt":

```Python
for text in msp.query("TEXT"):
    style_name = text.dxf.style
    style = doc.styles.get(style_name)
    print(style.dxf.font)
    # same as:
    print(text.font_name())

```

To render MTEXT entities, explode them by the `MTextExplode` add-on into TEXT entities.

The method `text.text_transformation_matrix()`  returns the `Matrix44` instance to transform the `text_path` to the WCS location of the TEXT entity:

```Python
wcs_path = text_path.transform(text.text_transformation_matrix())
```

**Comment by typhoon71** [comment]
That's exactly what I needed.
Good to know you're adding support for .lff fonts too, but with this I can already do what I needed.
An exaustive answer, very useful, thank you.

**Comment by typhoon71** [comment]
Just a little update for version 1.1.0:

one import changes from:
```python
from ezdxf import shapefile
```
to:

```python
from ezdxf.fonts import shapefile
```
and this code from the previous examples:

```python
text_path = font.render_text(text)
path.render_splines_and_polylines(msp, [text_path])
```

to:

```python 
text_path = font.render_text(text).to_path()
path.render_splines_and_polylines(msp, [text_path])
```

Updating this since I couldn't find anything about the second one in the docs and this use is pretty uncommon.

**Reply by mozman** [reply]
You are correct! But ... 😄

Now that shape file support is built into the `fonts` subpackage, the code can be further simplified:

```Python
import ezdxf
from ezdxf import path, zoom
from ezdxf.fonts import fonts
from ezdxf.document import Drawing

def render_txt(fontname: str, text: str) -> Drawing:
    doc = ezdxf.new()
    msp = doc.modelspace()
    font = fonts.make_font(fontname, cap_height=3.5)
    text_path = font.text_path(text)
    # convert the optimized 2D text_path to a regular Path()
    path.render_splines_and_polylines(msp, [text_path.to_path()])
    return doc


doc = render_txt("iso.shx", "7410/A")
zoom.extents(doc.modelspace())
doc.saveas("iso_shx.dxf")
```

Shape file fonts and TTF fonts can be treated the same!

**Comment by typhoon71** [comment]
ehm, I tested the code in the last example you posted and... the output is not what I need.

I get:

![220989073-a98e554d-473e-4092-b067-1c9057161ac8](https://github.com/mozman/ezdxf/assets/1866248/7eda104c-dc0a-456c-94d9-b19f86b9a86a)

I need:

![220989014-493af6bf-5829-41f5-954c-23d89ebaca07](https://github.com/mozman/ezdxf/assets/1866248/9a37009c-1129-4701-8913-66cdbc518885)

Which is what I was getting when I opened this discussion; I guessI'm going to use:

```python 
font = shapefile.readfile(fontname)
```

**Reply by mozman** [reply]
Hi,

points to check:
- add the directory of the shx-fonts to the support directories: https://ezdxf.mozman.at/docs/tools/fonts.html#basic-stroke-fonts
- rebuild the font cache: https://ezdxf.mozman.at/docs/howto/fonts.html
- is the font really available

EDIT:

Adding the shx directory in the script does not work for `fonts.make_fonts()`, because the cache is loaded at startup. You need a config file for this solution.

**Comment by typhoon71** [comment]
mmm, too much work, postponed.

Scrap that, seems like I'm having issues with scaling. Will complete this when less sleepy.

**Reply by mozman** [reply]
The `ShapeFile.render_text()` method is a low-level interface and `make_font()` is a high-level interface. The result is the same. A high-level interface is more future-proof.

**Comment by typhoon71** [comment]
In a less sleep deprived state, I found what I was doing wrong.

I used to use:
```python
font = shapefile.readfile(fontpath)  # this was the actual full path of the font
```

But now: 
```python
font = fonts.make_font(fontname, cap_height=3.5)  # this is just the font name
```

As you suggested I had to add:
```python
ezdxf.options.support_dirs = [fontdir]  # my custom font path list
```
to have the desired font loaded; turns out I was misinterpreting what the class was doing.

It simplifies some of my code, getting cap_height from entity.dxf.height, along the other stuff I needed for insert.
Beautiful.
